documentation/api/query/v4/upgrading-from-v3.markdown

```diff
@@ -0,0 +1,125 @@
+  under the `value` field in the case of a structured fact. This is observable
+  in current stable (PDB 2.2.x or PE 3.7) with the v4 API.
+
+* The v4 catalogs endpoint no longer produces a `metadata` field or the
+  `api_version` key within, and the contents of the former `data` field now
+  compose the response body. The body contains the new fields
+  `producer_timestamp`, `hash`, and `environment`, and the endpoint is
+  documented here:
+  https://docs.puppetlabs.com/puppetdb/master/api/query/v4/catalogs.html
+
+* The only valid command submission versions for catalogs, reports, and facts
+  at release time will be 6, 5, and 4. These are respectively documented at
+  https://docs.puppetlabs.com/puppetdb/master/api/wire_format/catalog_format_v6.html,
+  https://docs.puppetlabs.com/puppetdb/master/api/wire_format/report_format_v5.html,
+  and
+  https://docs.puppetlabs.com/puppetdb/master/api/wire_format/facts_format_v4.html.
+
+* PuppetDB's mbeans (listed at /metrics/v1/mbeans) are no longer prefixed with
+  "com."
+
+### NEW ENDPOINTS
+
+* `/v4/factsets` The factsets endpoint did not exist in v3, but is present in
+  PuppetDB 2.2.x on the v4 API. PDB 3.0 will add a hash in the metadata.
+  Documentation is here:
+  https://docs.puppetlabs.com/puppetdb/master/api/query/v4/factsets.html#response-format
+
+* `/v4/fact-paths` This endpoint is similar to the existing fact-names endpoint
+  in that one expected use is GUI autocompletion. Then endpoint is documented
+  here:
+  https://docs.puppetlabs.com/puppetdb/master/api/query/v4/fact-paths.html
+
+* `/v4/fact-contents` This endpoint allows fine-grained querying of facts and
+  structured facts. Documentation here:
+  https://docs.puppetlabs.com/puppetdb/master/api/query/v4/fact-contents.html
+
+### NEW API FEATURES
+
+* The response of the reports endpoint includes the new fields `noop`,
+  `environment`, `status`, `resource_events`, `logs`, and `metrics`.  The
+  endpoint in its latest form is documented here:
+  https://docs.puppetlabs.com/puppetdb/master/api/query/v4/reports.html and an
+  example of the new format is here:
+  https://docs.puppetlabs.com/puppetdb/master/api/query/v4/reports.html#examples.
+  For comparison, the old response is here:
+  https://docs.puppetlabs.com/puppetdb/latest/api/query/v3/reports.html#response-format
+
+* The reports endpoint takes a latest_report? query to return only reports
+  associated with the most recent puppet run for their nodes. Similar to the
+  corresponding events query, there is no corresponding field in the response.
+  The report query fields are documented here:
+  https://docs.puppetlabs.com/puppetdb/master/api/query/v4/reports.html#query-fields
+
+* The v4 catalogs endpoint is queryable like the other endpoints, whereas
+  before it could only return a catalog for a single host. The old query format
+  (/v4/catalogs/myhost) still works as before, but /v4/catalogs returns results
```

Comment:
I don't remember where we landed on this, does /v4/catalogs/myhost return a single element or an array of 1?